workbench-core/datasets/src/dataSetService.ts

```diff
@@ -252,32 +258,54 @@ export class DataSetService {
    *
    * @returns an array of DataSet objects.
    */
-  public async listDataSets(authenticatedUser: AuthenticatedUser): Promise<DataSet[]> {
+  public async listDataSets(
+    authenticatedUser: AuthenticatedUser,
+    pageSize: number,
+    paginationToken: string | undefined
+  ): Promise<PaginatedResponse<DataSet>> {
     const metadata: Metadata = {
       actor: authenticatedUser,
       action: this.listDataSets.name,
       source: {
         serviceName: DataSetService.name
       }
     };
-
     try {
-      const response: DataSet[] = [];
-      const allDatasets = await this._dbProvider.listDataSets();
-
-      await Promise.all(
-        allDatasets.map(async (dataset) => {
-          try {
-            // throws a ForbiddenError if the user doesnt have permission on the dataset
-            await this._authzPlugin.isAuthorizedOnDataSet(dataset.id, 'read-only', authenticatedUser);
-            response.push(dataset);
-          } catch (error) {
-            if (!isForbiddenError(error)) {
-              throw error;
+      const authedDataSets: DataSet[] = [];
+
+      let lastPaginationToken = paginationToken;
+      do {
+        const pluginResponse = await this._dbProvider.listDataSets(pageSize, lastPaginationToken);
+
+        await Promise.all(
+          pluginResponse.data.map(async (dataset) => {
+            try {
+              await this._authzPlugin.isAuthorizedOnDataSet(dataset.id, 'read-only', authenticatedUser);
+              authedDataSets.push(dataset);
+            } catch (error) {
+              if (!isForbiddenError(error)) {
+                throw error;
+              }
             }
-          }
-        })
-      );
+          })
+        );
+        lastPaginationToken = pluginResponse.paginationToken;
+      } while (authedDataSets.length < pageSize && lastPaginationToken);
+
+      let sortedDataSets = authedDataSets.sort((a: DataSet, b: DataSet) => {
+        return Date.parse(b.createdAt) - Date.parse(a.createdAt);
+      });
+
+      if (authedDataSets.length > pageSize) {
+        sortedDataSets = sortedDataSets.slice(0, pageSize);
+        const lastKey = buildDynamoDBPkSk(sortedDataSets[pageSize - 1].id, resourceTypeToKey.dataset);
```

Comment:
Don't do this here; since this is ddb specific logic, it should be moved to one of the ddb plugins.

Reply:
Oh, I see what you mean. We need to expose a method in the DataSetsMetadataPlugin to do this! Good catch!